Moonscraper Chart Editor/Assets/Scripts/Game/Misc/SongValidate.cs

```diff
@@ -194,6 +194,50 @@ static string CheckForErrorsCloneHero(Song song, ValidationParameters validation
             }
         }
 
+        foreach (var instrument in EnumX<Song.Instrument>.Values)
+        {
+            if (instrument != Song.Instrument.Drums && instrument != Song.Instrument.Unrecognised)
+            {
+                Chart lastChart = null;
+                foreach (var difficulty in EnumX<Song.Difficulty>.Values)
+                {
+                    var chart = song.GetChart(instrument, difficulty);
+
+                    // Skip if chart is same as last
+                    if (chart == lastChart)
+                    {
+                        continue;
+                    }
+
+                    lastChart = chart;
+
+                    foreach (var note in chart.notes)
+                    {
+                        bool previousSameTick = note.previous != null && note.tick == note.previous.tick;
+                        bool nextSameTick = note.next != null && note.tick == note.next.tick;
+
+                        // Open chords are not supported in Clone Hero (yet)
+                        if (note.IsOpenNote() && (previousSameTick || nextSameTick))
+                        {
+                            sb.AppendFormat("\tFound Open chord at time {1}, position {0}.\n",
+                                note.tick, PrintObjectTime(note.time));
+
+                            hasErrorsLocal |= true;
+                        }
+
+                        // Neither are Tap opens
+                        if (note.IsOpenNote() && (note.flags & Note.Flags.Tap) != 0)
+                        {
+                            sb.AppendFormat("\tFound Tap Open note at time {1}, position {0}.\n",
+                                note.tick, PrintObjectTime(note.time));
+
+                            hasErrorsLocal |= true;
+                        }
+                    }
```

Comment:
Would you be able to gather some performance metrics on this? By default the song validator will run every single save, so this could cause a bad hitch on larger charts. Would like to get some timing numbers on something like Crash Test 5 for example.

Reply:
So the performance metrics I gathered is that it never completes because Unity begins allocating so much memory that it runs out and crashes. I'm going to guess that's to do with the StringBuilder, so I'll test it without that

Reply:
Turns out its nothing to do with the open note validator, but rather the MS validator for out of range objects.

![image](https://github.com/user-attachments/assets/cc766790-6918-45dc-8443-821b922c7bac)

![image](https://github.com/user-attachments/assets/0c18f330-06f1-477a-9000-783934d22cce)

The actual issue is not the StringBuilder either, as this would only allocate around ~400MB of memory. I believe the actual issue comes from the Scroll Rect for the error text box, as when Unity crashes it is trying to allocate an array in the LateUpdate call for Scroll Rect.

If I select the audio for Crash Test 5 so none of the objects are out of range of the song's length, all the validation runs fine.

Reply:
With the MS validator disabled so I can actually profile my own code, this is the result for the Open Chord & Tap validator that I added:

![image](https://github.com/user-attachments/assets/e40747cf-ad89-46d7-a933-a48d71e884d3)

Granted Crash Test 5 has none of these features, but the overhead of checking the ticks, tap flag then StringBuilder append would be minuscule in comparison to the 1+ million notes to iterate.

Reply:
Hm, so there would be a notable hitch upon saving. Probably fine tbh considering how much of an outlier CT5 is, but it worth noting that some work to rework the song validator run asynchronously may be needed.